client/comms/wsmanager.go

```diff
@@ -0,0 +1,200 @@
+	mgr.reqMtx.Lock()
+	mgr.req[id] = route
+	mgr.reqMtx.Unlock()
+}
+
+// fetchRoute fetches the route of the provided request id.
+func (mgr *WsManager) fetchRoute(id uint64) (string, error) {
+	mgr.reqMtx.RLock()
+	route := mgr.req[id]
+	mgr.reqMtx.RUnlock()
+
+	if route == "" {
+		return "", fmt.Errorf("no route found for id #%v", id)
+	}
+
+	return route, nil
+}
+
+// NextID returns the next request id.
+func (mgr *WsManager) NextID() uint64 {
+	return atomic.AddUint64(&mgr.rID, 1)
+}
+
+// connect attempts to establish a websocket connection.
+func (mgr *WsManager) connect() error {
+	url := url.URL{
+		Scheme: "ws",
+		Host:   mgr.host,
+	}
+
+	var err error
+	mgr.ws, _, err = websocket.DefaultDialer.Dial(url.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	// Set the pong hander to atomically update the last pong time.
+	mgr.ws.SetPongHandler(func(data string) error {
+		now := time.Now().Unix()
+		atomic.StoreInt64(&mgr.lastPong, now)
+		return mgr.ws.SetReadDeadline(time.Now().Add(time.Second * pingWait))
+	})
+
+	return nil
+}
+
+// keepAlive maintains an active websocket connection by monitoring pongs to
+// the server and reconnecting if the established connection is broken.
+// This should be run as a goroutine.
+func (mgr *WsManager) keepAlive(ctx context.Context) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			now := time.Now().Unix()
+			lastPong := atomic.LoadInt64(&mgr.lastPong)
+
+			// Reconnect the websocket when the ping wait threshold is exceeded.
+			if now-lastPong > pingWait {
+				atomic.StoreInt64(&mgr.lastPong, 0)
+				err := mgr.connect()
+				if err != nil {
+					log.Errorf("unable to create websocket "+
+						"connection: %v", err)
+				}
+			}
+
+		case <-ctx.Done():
+			mgr.wg.Done()
+			return
+		}
+	}
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (mgr *WsManager) read() {
+	for {
+		_, data, err := mgr.ws.ReadMessage()
+		if err != nil {
+			if err != io.EOF {
+				log.Errorf("unable to read message: %v", err)
+				continue
+			}
+
+			return
+		}
+
+		log.Tracef("read message: %v", spew.Sdump(data))
+
+		msg := new(msgjson.Message)
+		if err != json.Unmarshal(data, msg) {
+			log.Errorf("unable to parse message %v: %v", string(data), err)
+			continue
+		}
+
+		mgr.readCh <- msg
+	}
+}
+
+// send pushes outgoing messages over the websocket connection.
+// This should be run as a goroutine.
+func (mgr *WsManager) send(ctx context.Context) {
+	for {
+		select {
+		case msg := <-mgr.sendCh:
+			log.Tracef("sending message: %v", spew.Sdump(msg))
+
+			err := mgr.ws.WriteJSON(msg)
+			if err != nil {
+				log.Errorf("unable to write message: %v", err)
+				continue
+			}
+
+		case <-ctx.Done():
+			mgr.wg.Done()
+			return
+		}
+	}
+}
+
+// run starts all processes of the websocket manager.
+// This should be run as a goroutine.
+func (mgr *WsManager) run(ctx context.Context) {
```

Comment:
Let's not run this as a goroutine. See https://github.com/decred/dcrdex/blob/2e850b0e0f56085b6fa61eeaf148d24ff0f10b97/server/comms/link.go#L176-L180